File: leetcode/leetcode_py/implement-trie-prefix-tree.py
'''
https://leetcode.com/problems/implement-trie-prefix-tree

Solution. Build Tries in place (array, hashmap). 
time
- insert: o(n)
- search: o(n)
- startWith: o(n)
space: o(n)
'''
class Trie:

    def __init__(self):
        self.children = {}
        self.isLeaf = False

    def insert(self, word: str) -> None:
        node = self
        for c in word:
            if not c in node.children:
                node.children[c] = Trie()
            node = node.children[c]
        node.isLeaf = True
        
    def search(self, word: str) -> bool:
        node = self
        for c in word:
            if not c in node.children:
                return False
            node = node.children[c]
        return node.isLeaf        

    def startsWith(self, prefix: str) -> bool:
        node = self
        for c in prefix:
            if not c in node.children:
                return False
            node = node.children[c]
        return True        


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: leetcode/leetcode_cpp/implement-trie-prefix-tree.cpp
/*
https://leetcode.com/problems/implement-trie-prefix-tree

Solution. Build Tries in place (array, hashmap). 
time
- insert: o(n)
- search: o(n)
- startWith: o(n)
space: o(n)

*/
class Trie {
public:
    Trie() {        
    }
    
    void insert(string word) {
        Trie* node = this;
        for (auto c : word) {
            c -= 'a';
            if (!node->children[c]) node->children[c] = new Trie();
            node = node->children[c];
        }
        node->isLeaf = true;        
    }
    
    bool search(string word) {
        Trie* node = this;
        for (auto c : word) {
            c -= 'a';
            if (!node->children[c]) return false;
            node = node->children[c];
        }
        return node->isLeaf;
    }
    
    bool startsWith(string prefix) {
        Trie* node = this;
        for (auto c : prefix) {
            c -= 'a';
            if (!node->children[c]) return false;
            node = node->children[c];
        }
        return true;
    }

    Trie* children[26] = {};
    bool isLeaf = false;
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */